logs/ingress/src/audit_log.rs

```diff
@@ -0,0 +1,252 @@
+/// repeatedly searching the most recent audit log entries to find a matching
+/// entry.
+///
+/// Times out after 15 minutes
+pub async fn get_entry<P>(http: Arc<Http>, search: SearchQuery<P>) -> Result<AuditLogEntry>
+where
+    P: Fn(&AuditLogEntry) -> bool,
+{
+    let mut backoff = search.make_backoff();
+    let timing = SearchTiming {
+        start: time::millisecond_ts(),
+        target: search
+            .target_timestamp
+            .unwrap_or_else(|| time::millisecond_ts()),
+    };
+
+    loop {
+        let result = try_get_entry(Arc::clone(&http), &search, timing).await;
+        let err = match result {
+            Ok(v) => return Ok(v),
+            Err(err) => err,
+        };
+
+        match err {
+            BackoffError::Permanent(err) => return Err(err),
+            BackoffError::Transient(err) => {
+                if let Error::SerenityError(serenity_err) = &err {
+                    if let Some(response) = unauthorized_response(&serenity_err) {
+                        return Err(Error::Unauthorized(response));
+                    }
+                }
+            }
+        };
+
+        let next = match backoff.next_backoff() {
+            Some(next) => next,
+            None => return Err(Error::TimedOut),
+        };
+
+        tokio::time::delay_for(next).await;
+    }
+}
+
+/// Attempts to get a single audit log entry during a backoff loop
+async fn try_get_entry<P>(
+    http: Arc<Http>,
+    search: &SearchQuery<P>,
+    timing: SearchTiming,
+) -> std::result::Result<AuditLogEntry, BackoffError<Error>>
+where
+    P: Fn(&AuditLogEntry) -> bool,
+{
+    // limit is a tradeoff of time in situations where a single event of the target type
+    // occurs in a small interval to situations where many events of the target type
+    // occur in a small interval
+    let limit = search.chunk_size.unwrap_or(5);
+    // determines the max number of seconds to search back in the
+    // audit log for an entry since the target timestamp
+    let time_threshold = search
+        .timestamp_ignore_threshold
+        .map(|dur| dur.as_millis() as u64)
+        .unwrap_or(60_000);
+
+    let mut before: Option<AuditLogEntryId> = None;
+    // traverse the audit log history
+    loop {
+        let entries = search
+            .guild_id
+            .audit_logs(
+                Arc::clone(&http),
+                search.entry_type.map(|t| t as u8),
+                search.user_id,
+                before,
+                Some(limit),
+            )
+            .await
+            .map_err(|err| Error::SerenityError(err))?
+            .entries;
+
+        if entries.len() == 0 {
+            break;
+        }
+
+        // attempt to find the desired matching entry
+        let mut oldest: Option<AuditLogEntryId> = None;
+        for (key, value) in entries {
+            if (search.matches)(&value) && search.strategy.matches(timing, &value, search) {
```

Comment:
You don't need the parens around the function pointer.

Reply:
rustc yells at me that `matches` is a `field, not a method` :(

Reply:
Weird. Not sure why it doesn't work.

Reply:
You could maybe try a function pointer instead for your variadic type. Have `P: fn(&input) -> bool`

Reply:
Wouldn't that mean that I couldn't pass closures to the function? I'd rather keep the callsite interface as ergonomic as possible.

Reply:
Rust will autocast a closure to a function pointer. But this is mainly me just finding that syntax ugly and not an actual problem with the code so don't worry about it too much.